File: tests/stats_test.go
//go:build linux

package tests

import (
	"testing"
	"time"

	internal "nfs-gazer/internal"
)

func TestParseEvents(t *testing.T) {
	tests := []struct {
		name     string
		parts    []string
		expected *internal.NFSEvents
		wantErr  bool
	}{
		{
			name: "valid events with all fields",
			parts: []string{
				"1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
				"11", "12", "13", "14", "15", "16", "17", "18", "19", "20",
				"21", "22", "23", "24", "25", "26", "27",
			},
			expected: &internal.NFSEvents{
				InodeRevalidate:  1,
				DentryRevalidate: 2,
				DataInvalidate:   3,
				AttrInvalidate:   4,
				VFSOpen:          5,
				VFSLookup:        6,
				VFSAccess:        7,
				VFSUpdatePage:    8,
				VFSReadPage:      9,
				VFSReadPages:     10,
				VFSWritePage:     11,
				VFSWritePages:    12,
				VFSGetdents:      13,
				VFSSetattr:       14,
				VFSFlush:         15,
				VFSFsync:         16,
				VFSLock:          17,
				VFSRelease:       18,
				CongestionWait:   19,
				SetattrTrunc:     20,
				ExtendWrite:      21,
				SillyRename:      22,
				ShortRead:        23,
				ShortWrite:       24,
				Delay:            25,
				PNFSRead:         26,
				PNFSWrite:        27,
			},
			wantErr: false,
		},
		{
			name:     "insufficient parts",
			parts:    []string{"1", "2", "3"},
			expected: &internal.NFSEvents{},
			wantErr:  true,
		},
		{
			name: "minimum valid parts",
			parts: []string{
				"1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
				"11", "12", "13", "14", "15", "16", "17", "18", "19", "20",
				"21", "22", "23", "24", "25", "26", "27",
			},
			expected: &internal.NFSEvents{
				InodeRevalidate:  1,
				DentryRevalidate: 2,
				DataInvalidate:   3,
				AttrInvalidate:   4,
				VFSOpen:          5,
				VFSLookup:        6,
				VFSAccess:        7,
				VFSUpdatePage:    8,
				VFSReadPage:      9,
				VFSReadPages:     10,
				VFSWritePage:     11,
				VFSWritePages:    12,
				VFSGetdents:     13,
				VFSSetattr:      14,
				VFSFlush:        15,
				VFSFsync:        16,
				VFSLock:         17,
				VFSRelease:      18,
				CongestionWait:  19,
				SetattrTrunc:    20,
				ExtendWrite:     21,
				SillyRename:     22,
				ShortRead:       23,
				ShortWrite:      24,
				Delay:           25,
				PNFSRead:        26,
				PNFSWrite:       27,
			},
			wantErr: false,
		},
		{
			name: "invalid number format",
			parts: []string{
				"invalid", "2", "3", "4", "5", "6", "7", "8", "9", "10",
				"11", "12", "13", "14", "15", "16", "17", "18", "19", "20",
				"21", "22", "23", "24", "25", "26", "27",
			},
			expected: nil,
			wantErr:  true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := internal.ParseEvents(tt.parts)
			
			if tt.wantErr && err == nil {
				t.Errorf("internal.ParseEvents() expected error but got none")
				return
			}
			if !tt.wantErr && err != nil {
				t.Errorf("internal.ParseEvents() unexpected error: %v", err)
				return
			}
			
			if tt.expected == nil {
				if result != nil {
					t.Errorf("internal.ParseEvents() expected nil result but got: %v", result)
				}
				return
			}

			if result.InodeRevalidate != tt.expected.InodeRevalidate {
				t.Errorf("internal.ParseEvents() InodeRevalidate = %v, want %v", result.InodeRevalidate, tt.expected.InodeRevalidate)
			}
			if result.VFSOpen != tt.expected.VFSOpen {
				t.Errorf("internal.ParseEvents() VFSOpen = %v, want %v", result.VFSOpen, tt.expected.VFSOpen)
			}
			if result.PNFSRead != tt.expected.PNFSRead {
				t.Errorf("internal.ParseEvents() PNFSRead = %v, want %v", result.PNFSRead, tt.expected.PNFSRead)
			}
		})
	}
}

func TestCalculateDelta(t *testing.T) {
	tests := []struct {
		name     string
		oldOp    *internal.NFSOperation
		newOp    *internal.NFSOperation
		duration float64
		expected *internal.DeltaStats
	}{
		{
			name: "valid delta calculation",
			oldOp: &internal.NFSOperation{
				Name:        "READ",
				Ops:         100,
				BytesSent:   1000,
				BytesRecv:   5000,
				RTT:         500,
				ExecuteTime: 200,
				QueueTime:   100,
				Errors:      1,
				Timeouts:    2,
			},
			newOp: &internal.NFSOperation{
				Name:        "READ",
				Ops:         150,
				BytesSent:   1500,
				BytesRecv:   7500,
				RTT:         750,
				ExecuteTime: 300,
				QueueTime:   150,
				Errors:      3,
				Timeouts:    4,
			},
			duration: 1.0,
			expected: &internal.DeltaStats{
				Operation:    "READ",
				DeltaOps:     50,
				DeltaSent:    500,
				DeltaRecv:    2500,
				DeltaBytes:   3000,
				DeltaRTT:     250,
				DeltaExec:    100,
				DeltaQueue:   50,
				DeltaErrors:  2,
				DeltaRetrans: 2,
				IOPS:         50.0,
				AvgRTT:       5.0,
				AvgExec:      2.0,
				AvgQueue:     1.0,
				KBPerOp:      60.0 / 1024,
				KBPerSec:     3000.0 / 1024,
			},
		},
		{
			name:     "nil old operation",
			oldOp:    nil,
			newOp:    &internal.NFSOperation{Name: "READ", Ops: 50},
			duration: 1.0,
			expected: nil,
		},
		{
			name:     "nil new operation",
			oldOp:    &internal.NFSOperation{Name: "READ", Ops: 50},
			newOp:    nil,
			duration: 1.0,
			expected: nil,
		},
		{
			name: "no operation increase",
			oldOp: &internal.NFSOperation{
				Name: "READ",
				Ops:  100,
			},
			newOp: &internal.NFSOperation{
				Name: "READ",
				Ops:  100,
			},
			duration: 1.0,
			expected: &internal.DeltaStats{
				Operation: "READ",
				DeltaOps:  0,
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := internal.CalculateDelta(tt.oldOp, tt.newOp, tt.duration)
			
			if tt.expected == nil {
				if result != nil {
					t.Errorf("internal.CalculateDelta() expected nil but got: %v", result)
				}
				return
			}
			
			if result == nil {
				t.Errorf("internal.CalculateDelta() expected result but got nil")
				return
			}

			if result.Operation != tt.expected.Operation {
				t.Errorf("internal.CalculateDelta() Operation = %v, want %v", result.Operation, tt.expected.Operation)
			}
			if result.DeltaOps != tt.expected.DeltaOps {
				t.Errorf("internal.CalculateDelta() DeltaOps = %v, want %v", result.DeltaOps, tt.expected.DeltaOps)
			}
			if result.IOPS != tt.expected.IOPS {
				t.Errorf("internal.CalculateDelta() IOPS = %v, want %v", result.IOPS, tt.expected.IOPS)
			}
		})
	}
}

func TestParseMountstats(t *testing.T) {
	// Test that the function handles non-existent files gracefully
	t.Run("non-existent file", func(t *testing.T) {
		result, err := internal.ParseMountstats("/nonexistent/file")
		if err == nil {
			t.Errorf("internal.ParseMountstats() expected error for non-existent file")
		}
		if result != nil {
			t.Errorf("internal.ParseMountstats() expected nil result for non-existent file")
		}
	})
}


func TestDisplayStatsSimple(t *testing.T) {
	mount := &internal.NFSMount{
		Device:     "server:/export",
		MountPoint: "/mnt/nfs",
	}

	stats := []*internal.DeltaStats{
		{
			Operation: "READ",
			IOPS:      25.0,
			AvgRTT:    5.0,
			AvgExec:   3.0,
			KBPerSec:  100.0,
			KBPerOp:   2.0,
		},
	}

	// Test that the function doesn't panic
	t.Run("no panic on display", func(t *testing.T) {
		defer func() {
			if r := recover(); r != nil {
				t.Errorf("internal.DisplayStatsSimple() panicked: %v", r)
			}
		}()
		
		// Capture output would require redirecting stdout
		// For now, just ensure it doesn't crash
		internal.DisplayStatsSimple(mount, stats, false, time.Now())
	})
}

File: tests/stats_test.rs
use nfs_gaze::{parse_events, parse_nfs_operation, NFSEvents, NFSOperation};

#[test]
fn test_parse_events() {
    struct TestCase {
        name: &'static str,
        parts: Vec<&'static str>,
        want_err: bool,
        verify: Option<fn(&NFSEvents)>,
    }

    let tests = vec![
        TestCase {
            name: "valid events with all fields",
            parts: vec![
                "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15",
                "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27",
            ],
            want_err: false,
            verify: Some(|e: &NFSEvents| {
                assert_eq!(e.inode_revalidate, 1, "InodeRevalidate should be 1");
                assert_eq!(e.pnfs_write, 27, "PNFSWrite should be 27");
            }),
        },
        TestCase {
            name: "valid events without pNFS fields",
            parts: vec![
                "100", "200", "300", "400", "500", "600", "700", "800", "900", "1000", "1100",
                "1200", "1300", "1400", "1500", "1600", "1700", "1800", "1900", "2000", "2100",
                "2200", "2300", "2400", "2500",
            ],
            want_err: false, // This should now succeed with 25 parts
            verify: None,
        },
        TestCase {
            name: "insufficient parts",
            parts: vec!["1", "2", "3"],
            want_err: true,
            verify: None,
        },
        TestCase {
            name: "invalid number format",
            parts: vec![
                "a", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15",
                "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27",
            ],
            want_err: true,
            verify: None,
        },
    ];

    for test in tests {
        let parts: Vec<String> = test.parts.into_iter().map(|s| s.to_string()).collect();
        let result = parse_events(&parts);

        match (result, test.want_err) {
            (Ok(events), false) => {
                if let Some(verify) = test.verify {
                    verify(&events);
                }
            }
            (Err(_), true) => {
                // Expected error
            }
            (Ok(_), true) => {
                panic!("Test '{}': expected error but got success", test.name);
            }
            (Err(e), false) => {
                panic!("Test '{}': unexpected error: {}", test.name, e);
            }
        }
    }
}

#[test]
fn test_parse_nfs_operation() {
    struct TestCase {
        name: &'static str,
        op_name: &'static str,
        stats: Vec<&'static str>,
        want_err: bool,
        verify: Option<fn(&NFSOperation)>,
    }

    let tests = vec![
        TestCase {
            name: "valid operation with all fields",
            op_name: "READ",
            stats: vec!["100", "95", "5", "1024", "2048", "10", "20", "30", "2"],
            want_err: false,
            verify: Some(|op: &NFSOperation| {
                assert_eq!(op.name, "READ", "Name should be READ");
                assert_eq!(op.ops, 100, "Ops should be 100");
                assert_eq!(op.ntrans, 95, "Ntrans should be 95");
                assert_eq!(op.timeouts, 5, "Timeouts should be 5");
                assert_eq!(op.bytes_sent, 1024, "BytesSent should be 1024");
                assert_eq!(op.bytes_recv, 2048, "BytesRecv should be 2048");
                assert_eq!(op.queue_time, 10, "QueueTime should be 10");
                assert_eq!(op.rtt, 20, "RTT should be 20");
                assert_eq!(op.execute_time, 30, "ExecuteTime should be 30");
                assert_eq!(op.errors, 2, "Errors should be 2");
            }),
        },
        TestCase {
            name: "insufficient stats",
            op_name: "WRITE",
            stats: vec!["100", "95"],
            want_err: true,
            verify: None,
        },
        TestCase {
            name: "invalid number format",
            op_name: "READ",
            stats: vec!["abc", "95", "5", "1024", "2048", "10", "20", "30", "2"],
            want_err: true,
            verify: None,
        },
        TestCase {
            name: "zero values",
            op_name: "GETATTR",
            stats: vec!["0", "0", "0", "0", "0", "0", "0", "0", "0"],
            want_err: false,
            verify: Some(|op: &NFSOperation| {
                assert_eq!(op.name, "GETATTR", "Name should be GETATTR");
                assert_eq!(op.ops, 0, "Ops should be 0");
                assert_eq!(op.ntrans, 0, "Ntrans should be 0");
                assert_eq!(op.timeouts, 0, "Timeouts should be 0");
                assert_eq!(op.bytes_sent, 0, "BytesSent should be 0");
                assert_eq!(op.bytes_recv, 0, "BytesRecv should be 0");
                assert_eq!(op.queue_time, 0, "QueueTime should be 0");
                assert_eq!(op.rtt, 0, "RTT should be 0");
                assert_eq!(op.execute_time, 0, "ExecuteTime should be 0");
                assert_eq!(op.errors, 0, "Errors should be 0");
            }),
        },
    ];

    for test in tests {
        let stats: Vec<String> = test.stats.into_iter().map(|s| s.to_string()).collect();
        let result = parse_nfs_operation(test.op_name, &stats);

        match (result, test.want_err) {
            (Ok(operation), false) => {
                if let Some(verify) = test.verify {
                    verify(&operation);
                }
            }
            (Err(_), true) => {
                // Expected error
            }
            (Ok(_), true) => {
                panic!("Test '{}': expected error but got success", test.name);
            }
            (Err(e), false) => {
                panic!("Test '{}': unexpected error: {}", test.name, e);
            }
        }
    }
}
